src/crush/CrushWrapper.cc

```diff
@@ -2347,7 +2347,12 @@ int CrushWrapper::_choose_type_stack(
       underfull_buckets[j].insert(item);
     }
   }
-  ldout(cct, 20) << __func__ << " underfull_buckets " << underfull_buckets << dendl;
+  ldout(cct, 20) << __func__ << " underfull_buckets " << "[";
+  for (auto p = underfull_buckets.begin(); p != underfull_buckets.end(); ++p) {
+    if (p != underfull_buckets.begin()) *_dout << ",";
+    *_dout << *p;
+  }
+  *_dout << "]" << dendl;
 
```

Comment:
instead specializing `vector<set<int>>`, i think we should forward declare 
```c++
template<class A, class Comp, class Alloc>
inline ostream& operator<<(ostream&, const set<A, Comp, Alloc>&);
```
before the definition of 
```c++
template<class A, class Alloc>
inline ostream& operator<<(ostream& out, const vector<A,Alloc>& v)  ///...
```

then ADL will be able to find the definition `operator<<(ostream&, const set<A, Comp, Alloc>&)`.